File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int, int>> stk;
        int res = INT_MIN;

        for (int i = 0; i < heights.size(); i++) {
            int idx = i;
            while (!stk.empty() && heights[i] < stk.top().first) {
                pair<int, int> temp = stk.top();
                stk.pop();
                res = max(res, (i - temp.second)*temp.first);
                idx = temp.second;
            }
            stk.push({heights[i], idx});
        }
        int n = heights.size();
        while(!stk.empty()) {
            pair<int, int> tmp = stk.top();
            stk.pop();
            res = max(res, (n - tmp.second) * tmp.first);
        }
        return res;

    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:

        inc_stack = []
        res = 0

        for i, height in enumerate(heights):
            idx = i
            while inc_stack and height < inc_stack[-1][1]:
                j,ht = inc_stack.pop()
                res = max(res,ht * (i - j))
                idx = j

            inc_stack.append((idx,height))
        
        while inc_stack:
            idx,ht = inc_stack.pop()
            res = max(res, ht * (len(heights) - idx))
        
        return res
        

